src/mds/InoTable.cc

```diff
@@ -77,18 +103,49 @@ void InoTable::project_alloc_ids(interval_set<inodeno_t>& ids, int want)
 void InoTable::apply_alloc_ids(interval_set<inodeno_t>& ids)
 {
   dout(10) << "apply_alloc_ids " << ids << " to " << projected_free << "/" << free << dendl;
-  free.subtract(ids);
+  if (need_recycle) {
+    interval_set<inodeno_t> s;
+    s.intersection_of(recycle, projected_free);
+    if (!s.empty()) {
+      recycle.subtract(s);
+      free.insert(s);
```

Comment:
> Why not do this together in [1], I think it will make the code more readable ?
> 
> [1] [2d2f779#diff-b8fb49f247b586c5857057a870bb5c1446fb7c3fc5567ec3c144bac24593ca8dR83-R88](https://github.com/ceph/ceph/commit/2d2f7790ba6f5adff2fa06f6e83021528b9abea9#diff-b8fb49f247b586c5857057a870bb5c1446fb7c3fc5567ec3c144bac24593ca8dR83-R88) ?

I think all of the changes about `free` and `recycle` is only can be approved when journaling complete

Reply:
Since you have moved the all the `projected_recycle` to `project_free` there, so it should be okay to move the `recycle` if I didn't miss something important that we couldn't.

And then you can simplify the code here:

```
 void InoTable::apply_alloc_ids(interval_set<inodeno_t>& ids)
 {
   dout(10) << "apply_alloc_ids " << ids << " to " << projected_free << "/" << free << dendl;
    interval_set<inodeno_t> s;
    s.intersection_of(ids, free);
    if (!s.empty()) {
      free.subtract(s);
    }
    s.clear();
    s.intersection_of(ids, recycle);
    if (!s.empty()) {
      recycle.subtract(s);
    }
    ++version;
 }
```

Reply:
> Since you have moved the all the `projected_recycle` to `project_free` there, so it should be okay to move the `recycle` if I didn't miss something important that we couldn't.
> 
> And then you can simplify the code here:
> 
> ```
>  void InoTable::apply_alloc_ids(interval_set<inodeno_t>& ids)
>  {
>    dout(10) << "apply_alloc_ids " << ids << " to " << projected_free << "/" << free << dendl;
>     interval_set<inodeno_t> s;
>     s.intersection_of(ids, free);
>     if (!s.empty()) {
>       free.subtract(s);
>     }
>     s.clear();
>     s.intersection_of(ids, recycle);
>     if (!s.empty()) {
>       recycle.subtract(s);
>     }
>     ++version;
>  }
> ```

We also need to move unused inos from `recycle` to `free`